File: parser.cpp
#include <ctype.h>

#include "parser.hpp"

namespace ALisp {
	namespace Parser {
		namespace detail {
			char startOf(const TokenType &str) { return str[0]; }
			char endOf(const TokenType &str) { return *(str.cend() - 1); }
			TokenListType tokenise(const TokenType &str) EXCEPT {
				TokenListType tokens;
				const char *s = str.c_str();
				while (*s) {
					// Skip whitespace
					while (isspace(*s))
						++s;
					// Skip comment lines
					if (*s == ';' && *(s + 1) == ';')
						while (*s && *s != '\n' && *s != '\r')
							++s;
					// List open or close
					else if (*s == '(' || *s == ')')
						tokens.push_back(*s++ == '(' ? "(" : ")");
					// "String" in quotes
					else if (*s == '"') {
						const char *t = s;
						int escape = 0;
						do {
							++t;
							if (escape != 0) escape--;
							if (*t == '\\')
								escape = 2; // skip this and the next character
						} while (*t && (escape != 0 || *t != '"'));
						++t;
						tokens.push_back(StringType(s, t));
						s = t;
					// A generic token
					} else {
						const char *t = s;
						while (*t && !isspace(*t) && *t != '(' && *t != ')')
							++t;
						tokens.push_back(StringType(s, t));
						s = t;
					}
				}
				return tokens;
			}
			Cell atom(const TokenType &token) EXCEPT {
				if (startOf(token) == '"' && endOf(token) == '"')
					return StringCell(token.substr(1, token.length() - 2));
				if (isdigit(token[0]) || (token[0] == '-' && isdigit(token[1]))) {
					// Number
					auto isInteger = tryStringToNumber<IntegerType>(token);
					if (std::get<0>(isInteger))
						return IntCell(std::get<1>(isInteger));
					auto isFloat = tryStringToNumber<FloatType>(token);
					if (std::get<0>(isFloat))
						return FloatCell(std::get<1>(isFloat));
				}
				return Atoms::Declare(token.c_str());
			}
			Cell read_from(TokenListType &tokens) EXCEPT {
				const TokenType token(tokens.front());
				tokens.pop_front();
				if (token == "(") {
					ListCell cells;
					while (tokens.front() != ")") {
						cells.push(read_from(tokens));
					}
					tokens.pop_front();
					return cells;
				} else
					return atom(token);
			}
			Cell read(const StringType &str) EXCEPT {
				TokenListType tokens(tokenise(str));
				return read_from(tokens);
			}
		}
		Cell read(const StringType &str) {
			return detail::read(str);
		}
	}
}

File: lib/parser.js
/**
 * ALisp
 *  - A Lisp interpreter and platform in Node.js/Javascript
 *
 * parser - contains the ALisp parser
 */

"use strict";

var atoms = require('./atoms');
var errors = require('./errors');
var isx = require('./isx');
var isspace = isx.isspace, isdigit = isx.isdigit;
var types = require('./types');
var ALVar = types.ALVar, type = types.type;

function startOf (str) {
	if (str === '') return '';
	return str[0];
}

function endOf (str) {
	if (str === '') return '';
	return str[str.length - 1];
}

function tokenise(str) {
	var tokens = [];
	var s = 0, t; // string counters, use s for most and t when required
	while(s < str.length) {
		while (isspace(str[s]))                    // Skip whitespace
			++s;
		if (str.substr(s, 2) === ';;')             // Skip comment lines
			while(s < str.length && str[s] !== '\n' && str[s] !== '\r')
				++s;
		else if (str[s] === '(' || str[s] === ')') // List open or close
			tokens.push(str[s++] === '(' ? '(' : ')');
		else if (str[s] === '"') {                 // "string" in quotes
			t = s;
			var escape = 0;
			do {
				++t;
				if (escape !== 0) escape--;
				if (str[t] === '\\') escape = 2;
			} while (t < str.length && (escape !== 0 || str[t] !== '"'));
			++t;
			tokens.push(str.substr(s, t - s));
			s = t;
		} else {                                   // A generic token
			t = s;
			while (t < str.length && !isspace(str[t]) && str[t] !== '(' && str[t] !== ')')
				++t;
			tokens.push(str.substr(s, t - s));
			s = t;
		}
	}
	return tokens;
}

function atom (token) {
	if (startOf(token) === '"' && endOf(token) === '"')
		return new ALVar(token.substr(1, token.length - 2), type.STRING);
	if (isdigit(token[0]) || token.length > 1 && token[0] === '-' && isdigit(token[1])) {
		// coerce token to a number by calling Number class
		return new ALVar(Number(token), type.NUMBER);
	}
	return atoms.declare(token);
}

function read_from(tokens) {
	if (tokens.length === 0) throw new errors.ParserError("Missing opening token");

	var token = tokens.shift();
	if (token === "(") {
		var cells = new ALVar([], type.LIST);
		while (tokens[0] !== ")") {
			cells.push(read_from(tokens));
			if (tokens.length === 0) throw new errors.ParserError("Missing closing )");
		}
		if (tokens.length === 0) throw new errors.ParserError("Missing closing )");
		tokens.shift(); // discard closing )
		return cells;
	}
	return atom(token);
}

function read(str) {
	var tokens = tokenise(str);
	return read_from(tokens);
}

exports.read = read;
exports.detail = {
	atom: atom,
	read_from: read_from,
	tokenise: tokenise
};
